nuxeo-core/nuxeo-core-cache/src/main/java/org/nuxeo/ecm/core/transientstore/keyvalueblob/KeyValueBlobTransientStore.java

```diff
@@ -0,0 +1,572 @@
+            String json = kvs.getString(key + DOT_PARAMINFO);
+            List<String> parameters = jsonToList(json);
+            if (parameters == null) {
+                parameters = new ArrayList<>();
+            }
+            parameters.add(parameter);
+            String newJson;
+            try {
+                newJson = mapper.writeValueAsString(parameters);
+            } catch (IOException e) {
+                throw new NuxeoException(e);
+            }
+            if (kvs.compareAndSet(key + DOT_PARAMINFO, json, newJson, ttl)) {
+                break;
+            }
+        }
+        markEntryExists(key);
+    }
+
+    @Override
+    public Serializable getParameter(String key, String parameter) {
+        KeyValueStore kvs = getKeyValueStore();
+        String k = key + DOT_PARAM_DOT + parameter;
+        String format = kvs.getString(k + FORMAT);
+        if (format == null) {
+            return kvs.getString(k);
+        } else {
+            byte[] bytes = kvs.get(k);
+            return SerializationUtils.deserialize(bytes);
+        }
+    }
+
+    @Override
+    public void putParameters(String key, Map<String, Serializable> parameters) {
+        parameters.forEach((param, value) -> putParameter(key, param, value));
+    }
+
+    @Override
+    public Map<String, Serializable> getParameters(String key) {
+        KeyValueStore kvs = getKeyValueStore();
+        // get the list of keys
+        String json = kvs.getString(key + DOT_PARAMINFO);
+        List<String> parameters = jsonToList(json);
+        if (parameters == null) {
+            // if the entry doesn't exist at all return null, otherwise empty
+            if (kvs.getString(key + DOT_COMPLETED) == null) {
+                return null;
+            } else {
+                return Collections.emptyMap();
+            }
+        }
+        // get values
+        return parameters.stream().collect(Collectors.toMap(identity(), p -> getParameter(key, p)));
+    }
+
+    protected void removeParameters(String key) {
+        KeyValueStore kvs = getKeyValueStore();
+        String json = kvs.getString(key + DOT_PARAMINFO);
+        List<String> parameters = jsonToList(json);
+        if (parameters != null) {
+            for (String parameter : parameters) {
+                String k = key + DOT_PARAM_DOT + parameter;
+                kvs.put(k, (String) null);
+                kvs.put(k + FORMAT, (String) null);
+            }
+        }
+        kvs.put(key + DOT_PARAMINFO, (String) null);
+    }
+
+    @Override
+    public void putBlobs(String key, List<Blob> blobs) {
+        if (absoluteMaxSize > 0 && getStorageSize() > absoluteMaxSize) {
+            // do the costly computation of the exact storage size if needed
+            doGC();
+            if (getStorageSize() > absoluteMaxSize) {
+                throw new MaximumTransientSpaceExceeded();
+            }
+        }
+
+        // remove previous blobs
+        removeBlobs(key);
```

Comment:
Should we do that before checking the storage size ?

Reply:
Yes good point.

Reply:
(Although in practice in our current usage we never replace blobs once created.)

Reply:
Ok I kept it as is as it's the same behavior done in other transient stores. We'll clean them up all at the same time I think.